### ISSUE TITLE
[Wotlk] MMaps and Towers

### ISSUE THREAD

**Issue by Grz3s** [cmangos/issues#838]
Hi
Strange problem when MMaps are = "on"  in Towers:
- Tower of Klaven MortWake in Westfall
- Tower of Azora - Elvynn forest
- Tower of Ilgalar - Redridge Mountains
  and maybe more ... somewhere... (im sure all same type have same problem)

Something is blocking passage... that non of creatures can go through. (player can)
Here where it is:

![gpsa](https://cloud.githubusercontent.com/assets/2374069/12011137/b7931f10-acba-11e5-890b-1d34e7d531ef.jpg)

Also pls look at this avi: to see in detail what exact happens:
https://youtu.be/cyU83iUROu4

If mmaps are "OFF" ... creatures will walk fine there...

@cyberium (or anyone who can help) can You pls look at this ... using your... RecastDemo
http://cmangos.net/thread-88-post-41309.html#pid41309
Maybe you will find what is that magic barrier for creatures there...
Im not good enough (this tool wont even start for me - or im missing smth or i have no idea how to run it:) 
Thank you.

P.S
Itt may be problem for other ver. [classic] and [TBC] -- someone need to confirm

**Comment by Nightfly510** [comment]
This issue is caused by the _**WorldObject::UpdateAllowedPositionZ()**_ adjusting the Z-position excessively and the unit not being able to find a path to the (incorrect) position.  It also exists in Classic, and likely other versions.

I found an easy workaround, which remedies pet pathing like this example as well as quirky movement of creatures in caves evading/bugging/plotting strange paths in certain places.  Simply limit the amount that this call is allowed to change the Z-position (pretty much all the problems I've seen occur with the function returning too high of a value).  A limit of 5.0 has worked well in my experimentation.

I'm a newb with Git, so I'm not sure how to actually post a diff, but briefly:  Object.cpp, line 1273-1274 of the current build, you want to add another check.  Change:
```
1273          else if (z < ground_z)
1274               z = ground_z;
```

to
```
1273          else if (z < ground_z)
1274               if (ground_z - z <= 5.0)
1275                  z = ground_z;
```
This effectively bypasses the modification of Z if the change would be too great (greater than 5.0 in this case), and simply uses the original Z value.

**Comment by Nightfly510** [comment]
Ok, actually, I misinterpreted your original post - thinking you were primarily referencing pet pathing.  Watched your vid and seems you are talking about creature attacks.

The fix I proposed does still help with this (they will attack across that boundary), but I agree there is still something quirky with that area:  if I run across it they will path back down to the ground floor and back up the other ramp to get to where I move to.  My guess is for some reason it doesn't efficiently handle the case where there are two distinctly different paths to the same spot.  Or, perhaps as you suggest, there is some strange barrier.
